stonesoup/detector/beamformers.py

```diff
@@ -0,0 +1,544 @@
+
+        if (p_K > 1):
+            for sensor_ind in range(0, 9):
+                for k1 in range(0, p_K):
+                    for k2 in range(k1+1, p_K):
+                        DTD[k1, k2] = DTD[k1, k2] \
+                            + cosalpha[k1, sensor_ind] * cosalpha[k2, sensor_ind] * sumsinsq \
+                            + (cosalpha[k1, sensor_ind] * sinalpha[k2, sensor_ind]
+                               + cosalpha[k2, sensor_ind] * sinalpha[k1, sensor_ind]) * sumsincos \
+                            + sinalpha[k1, sensor_ind]*sinalpha[k2, sensor_ind] * sumcossq
+                        DTD[k2, k1] = DTD[k1, k2]
+
+        Dterm = np.matmul(np.linalg.solve(1001*DTD, DTy), np.transpose(DTy))
+        log_posterior = - (p_K * np.log(1.001) / 2) - (N / 2) * np.log((yTy - Dterm) / 2) \
+            + p_K * np.log(Lambda) - np.log(np.math.factorial(p_K)) - p_K*np.log(math.pi * math.pi)
+        # note: math.pi*math.pi comes from area of parameter space in one dimension (i.e. range of
+        # azimuth * range of elevation)
+
+        return log_posterior
+
+    def elevation_mode_coin_toss(self, p_params, p_K):
+        # elevation angle has refelctions / rotations of state space resulting in duplicate modes
+        # use a coin toss to decide whether we jump to another mode to ensure full exploration and
+        # mixing can combine mirrored modes in post processing
+        for k in range(0, p_K):
+            # transform to first mode
+            if p_params[k, 0] > 3*math.pi/2:
+                p_params[k, 0] = 2*math.pi - p_params[k, 0]
+                p_params[k, 1] = p_params[k, 1] - math.pi
+            elif p_params[k, 0] > math.pi:
+                p_params[k, 0] = p_params[k, 0] - math.pi
+                p_params[k, 1] = p_params[k, 1] - math.pi
+            elif p_params[k, 0] > math.pi/2:
+                p_params[k, 0] = math.pi - p_params[k, 0]
+
+            # do coin toss to decide mode to jump to
+            toss = random.uniform(0, 1)
+            if toss < 0.25:
+                # first mode
+                pass
+            elif toss < 0.5:
+                # second mode
+                p_params[k, 0] = math.pi - p_params[k, 0]
+            elif toss < 0.75:
+                # third mode
+                p_params[k, 0] = p_params[k, 0] + math.pi
+                p_params[k, 1] = p_params[k, 1] + math.pi
+            else:
+                # fourth mode
+                p_params[k, 0] = 2*math.pi - p_params[k, 0]
+                p_params[k, 1] = p_params[k, 1] + math.pi
+
+            # wrap angles
+            if p_params[k, 0] < 0:
+                p_params[k, 0] = p_params[k, 0] + 2*math.pi
+            elif p_params[k, 0] > 2*math.pi:
+                p_params[k, 0] = p_params[k, 0] - 2*math.pi
+            if p_params[k, 1] < 0:
+                p_params[k, 1] = p_params[k, 1] + 2*math.pi
+            elif p_params[k, 1] > 2*math.pi:
+                p_params[k, 1] = p_params[k, 1] - 2*math.pi
```

Comment:
There are `Bearing` and `Elevation` types that could be used for this.

Reply:
Thanks for this, I wasn't aware these existed. I've updated the parameter arrays to be StateVector[Elevation, Bearing]. However, I don't think I can use the `mod_angle` functions (at least for the elevation) as it's tied in with the MCMC algorithm, which works on the range [0, 2pi] for all angles due to the requirements of the method.